test/comp/markv_codec_test.cpp

```diff
@@ -0,0 +1,346 @@
+  TestEncodeDecode(R"(
+OpCapability Shader
+OpCapability Linkage
+OpCapability Int64
+OpMemoryModel Logical GLSL450
+%s64 = OpTypeInt 64 1
+%100 = OpConstant %s64 0
+%200 = OpConstant %s64 1
+%300 = OpConstant %s64 -1
+%400 = OpConstant %s64 9223372036854775807
+%500 = OpConstant %s64 -9223372036854775808
+)");
+}
+
+TEST(Markv, U16Literal) {
+  TestEncodeDecode(R"(
+OpCapability Shader
+OpCapability Linkage
+OpCapability Int16
+OpMemoryModel Logical GLSL450
+%u16 = OpTypeInt 16 0
+%100 = OpConstant %u16 0
+%200 = OpConstant %u16 1
+%300 = OpConstant %u16 65535
+)");
+}
+
+TEST(Markv, S16Literal) {
+  TestEncodeDecode(R"(
+OpCapability Shader
+OpCapability Linkage
+OpCapability Int16
+OpMemoryModel Logical GLSL450
+%s16 = OpTypeInt 16 1
+%100 = OpConstant %s16 0
+%200 = OpConstant %s16 1
+%300 = OpConstant %s16 -1
+%400 = OpConstant %s16 32767
+%500 = OpConstant %s16 -32768
+)");
+}
+
+TEST(Markv, F32Literal) {
+  TestEncodeDecode(R"(
+OpCapability Shader
+OpCapability Linkage
+OpMemoryModel Logical GLSL450
+%f32 = OpTypeFloat 32
+%100 = OpConstant %f32 0
+%200 = OpConstant %f32 1
+%300 = OpConstant %f32 0.1
+%400 = OpConstant %f32 -0.1
+)");
+}
+
+TEST(Markv, F64Literal) {
+  TestEncodeDecode(R"(
+OpCapability Shader
+OpCapability Linkage
+OpCapability Float64
+OpMemoryModel Logical GLSL450
+%f64 = OpTypeFloat 64
+%100 = OpConstant %f64 0
+%200 = OpConstant %f64 1
+%300 = OpConstant %f64 0.1
+%400 = OpConstant %f64 -0.1
+)");
+}
+
+TEST(Markv, F16Literal) {
+  TestEncodeDecode(R"(
+OpCapability Shader
+OpCapability Linkage
+OpCapability Float16
+OpMemoryModel Logical GLSL450
+%f16 = OpTypeFloat 16
+%100 = OpConstant %f16 0
+%200 = OpConstant %f16 1
+%300 = OpConstant %f16 0.1
+%400 = OpConstant %f16 -0.1
+)");
+}
+
+TEST(Markv, StringLiteral) {
+  TestEncodeDecode(R"(
+OpCapability Shader
+OpCapability Linkage
+OpExtension "SPV_KHR_16bit_storage"
+OpExtension "xxx"
+OpExtension "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+OpExtension ""
+OpMemoryModel Logical GLSL450
+)");
+}
+
+TEST(Markv, WithFunction) {
+  TestEncodeDecode(R"(
+OpCapability Addresses
+OpCapability Kernel
+OpCapability GenericPointer
+OpCapability Linkage
+OpExtension "SPV_KHR_16bit_storage"
+OpMemoryModel Physical32 OpenCL
+%f32 = OpTypeFloat 32
+%u32 = OpTypeInt 32 0
+%void = OpTypeVoid
+%void_func = OpTypeFunction %void
+%100 = OpConstant %u32 1
+%200 = OpConstant %u32 2
+%main = OpFunction %void None %void_func
+%entry_main = OpLabel
+%300 = OpIAdd %u32 %100 %200
+OpReturn
+OpFunctionEnd
+)");
+}
+
+TEST(Markv, ForwardDeclaredId) {
+  TestEncodeDecode(R"(
+OpCapability Addresses
+OpCapability Kernel
+OpCapability GenericPointer
+OpCapability Linkage
+OpMemoryModel Physical32 OpenCL
+OpEntryPoint Kernel %1 "simple_kernel"
+%2 = OpTypeInt 32 0
+%3 = OpTypeVector %2 2
+%4 = OpConstant %2 2
+%5 = OpTypeArray %2 %4
+%6 = OpTypeVoid
+%7 = OpTypeFunction %6
+%1 = OpFunction %6 None %7
+%8 = OpLabel
+OpReturn
+OpFunctionEnd
+)");
+}
+
```

Comment:
I'd like to see tests for:

- OpSwitch on a 64 bit value.  (proves that type tracking works somewhat)
- A test with instructions with a variable number of arguments (other than a function type), e.g. OpAccessChain
- A bit mask operand, e.g. OpLoopMerge with a combination of DontUnroll and DependencyLength 10.